rlclientlib/utility/apim_http_authorization.cc

```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include "apim_http_authorization.h"
+
+namespace reinforcement_learning {
+  int apim_http_authorization::init(const utility::configuration& config, api_status* status, i_trace* trace) {
+    _api_key = config.get(name::HTTP_API_KEY, "dummykey");
```

Comment:
If the API key is missing should this not be an error? We cannot proceed in this case

Reply:
Configuration never throw errors. This is consistent with the pattern through out. There is no change in behavior introduced in this PR.

Reply:
Just because a configuration hasn't thrown an error yet doesn't seem to be a valid reason to not do it here. Clearly the contract of this function is that it can fail, otherwise it would be a `void` function.

If the API key is not supplied and the `apim_http_authorization` object is in use is that an error? Can RLClientlib operate as normal in that case?

Reply:
@slahabar this issue is unresolved

Reply:
If the apim key is invalid, http_transport will report an error and the user has to handle it either in a) the background error handler or b) as warnings in foreground rank/report call.  This condition needs to be handled in any case since apim keys can be rotated and therefore become invalid.  However, from a user perspective, this is graceful system degradation since inference is still possible, even with bad apim keys and web pages are still being served etc.

Alternatively, treating this as an init failure means that rlclientlib itself will fail to initialize and the object will not be usable.  From an operational perspective, this requires catastrophic intervention.   i.e. shit has hit the fan and and I get a call because my website is down.

Given that this condition will be detected and handled for other reasons, this approach seems like an acceptable trade-off to me in this particular case.

That is not to say that missing config values cannot ever constitute a failure.

Reply:
It's a separate issue to have an expired or invalid key, and I agree because of rotation and many other things an invalid API key can sneak in here. I am referring to requiring a config value, valid or not. Just to be clear I am not proposing this function to "validate" an API key, just to check something was given.

I am just proposing the init failure is if **no** config value is supplied (not supplying an **invalid** value). Using a default API key of `dummykey` will never work. The fact you forgot to set a config value will be surfaced as a 403 in a background thread and is significantly harder to debug/diagnose than an immediate "this config field is required but not supplied".

If I were debugging rlclientlib I would prefer a foreground error stating "HTTP_API_KEY config required but was not supplied" instead of a background error stating "403 when trying to connect to APIM"

Reply:
Makes sense.  Should return an error here.

Reply:
Added check for api key being passed. Throws error if there is none.